Make force work in conditions for futures, not only promises.

raxoft opened the pull request:
The Future#**force** should return the resulting value, not the promise. Otherwise things like

``` ruby
   puts "true" if future{ true }.force
   puts "false" if future{ false }.force
   puts "nil" if future{ nil }.force
```

do not work as expected (currently all evaluate to true). I believe it should work consistently with promises:

``` ruby
   puts "true" if promise{ true }.force
   puts "false" if promise{ false }.force
   puts "nil" if promise{ nil }.force
```

The change above should achieve this. However note it may change how the marshalling works with regard to futures, as the special marshalling handling found in Promise becomes bypassed. You may want to review this aspect yourself.

raxoft committed:
Make force work in conditions for futures, not only promises.

The Future#__force__ should return the resulting value, not the promise. Otherwise things like
``` ruby
   puts "true" if future{ true }.force
   puts "false" if future{ false }.force
   puts "nil" if future{ nil }.force
```
do not work as expected (currently all evaluate to true). I believe it should work consistently with promises:
``` ruby
   puts "true" if promise{ true }.force
   puts "false" if promise{ false }.force
   puts "nil" if promise{ nil }.force
```
The change above should achieve this. However note it may change how the marshalling works with regard to futures, as the special marshalling handling found in Promise becomes bypassed. You may want to review this aspect yourself.

bhuga commented:
This seems valid at first glance. Can you create a test that demonstrates the problem?

raxoft commented:
The test case is the example above - it prints all three lines in case of futures, while only the first line for promises. Should be trivial to convert to the appropriate assert/expect/should/whatever.

I would need it for construct like this which I sometimes use for running things in parallel:

```
   something.map{ |x| future{ do_something_with x } }.map{ |x| x.force }
```

It works almost fine now, except that nil/false results are not properly handled in the resulting array, depending on how they are use.

bhuga commented:
> The test case is the example above

I'd need to see these added to the specs in the code. Note that travis is failing.

> something.map{ |x| future{ do_something_with x } }.map{ |x| x.force }

Can I ask why you're doing this? The point of future is that they start running immediately. The second `map` here isn't doing anything that the futures haven't already started doing as soon as they were created.

raxoft committed:
Added tests for results of futures and promises when used in conditions.

raxoft committed:
Marshaling really works for promises only, so adjust the specs accordingly.

raxoft commented:
I have added the tests to the specs. 

As for the travis failing, that's because the shared spec tests for compatibility with Marshal. However, Marshalling of futures doesn't work anyway (and didn't before), as Future class lacks the _load method and Marshal.load fails because of that. So I adjusted the spec accordingly. The alternative would be adding proper Marshalling support to Future class itself, basically cloning the functionality from Promise class.

As for the `map{ |x| x.force }` question: the first map runs some asynchronous requests which involve networking communication, the second map uses force to ensure that all the requests finish and any possible exceptions are raised at that point, within controlled context, before any further processing is done.

raxoft commented:
Hmm, the travis tests now pass for everything but JRuby, but checking the log it seems travis has problems running bundler under JRuby, rather than the tests themselves failing. I can't see how this can be fixed other than trying running the tests on travis again later.
